lib/wibox/layout/align.lua

```diff
@@ -41,121 +43,297 @@ local align = {}
 -- @param height The available height.
 function align:layout(context, width, height)
     local result = {}
-
-    -- Draw will have to deal with all three align modes and should work in a
-    -- way that makes sense if one or two of the widgets are missing (if they
-    -- are all missing, it won't draw anything.) It should also handle the case
-    -- where the fit something that isn't set to expand (for instance the
-    -- outside widgets when the expand mode is "inside" or any of the widgets
-    -- when the expand mode is "none" wants to take up more space than is
-    -- allowed.
-    local size_first = 0
-    -- start with all the space given by the parent, subtract as we go along
     local size_remains = self._private.dir == "y" and height or width
-    -- This is only set & used if expand ~= "inside" and we have second width.
-    -- It contains the size allocated to the second widget.
-    local size_second
-
-    -- we will prioritize the middle widget unless the expand mode is "inside"
-    --  if it is, we prioritize the first widget by not doing this block also,
-    --  if the second widget doesn't exist, we will prioritise the first one
-    --  instead
-    if self._private.expand ~= "inside" and self._private.second then
-        local w, h = base.fit_widget(self, context, self._private.second, width, height)
-        size_second = self._private.dir == "y" and h or w
-        -- if all the space is taken, skip the rest, and draw just the middle
-        -- widget
-        if size_second >= size_remains then
-            return { base.place_widget_at(self._private.second, 0, 0, width, height) }
-        else
-            -- the middle widget is sized first, the outside widgets are given
-            --  the remaining space if available we will draw later
-            size_remains = floor((size_remains - size_second) / 2)
+    local is_vert = self._private.dir == "y"
+
+    -- Get the required size of a child widget
+    local function fit_child(child, max_size)
+        -- Can't fit a widget if it does not exist
+        if not self._private[child] then
```

Comment:
I'm not a fan of these `self._private[child]` access. `child` can be anything, you have to use literal strings, and that's very error-prone.

Reply:
I tweaked how the functions work in https://github.com/awesomeWM/awesome/pull/3588/commits/d5575da0e53f65ab7c5ad55c3fad0f9e9eda6ebf). Now they accept the reference to the child widget instead of its name.